### ISSUE TITLE
Cannot use the same specification multiple times?

### ISSUE THREAD

**Issue by estelsmith** [K-Phoen/rulerz#3]
Hello. I love that someone is finally making a nice specification-pattern library for PHP. However, I have a concern I've ran across.

In the example below, it appears that if I want to filter by more than one group, it will overwrite the parameter with the last group that was specified. This effectively makes the specification as if I had set `$spec` to `new GroupSpec('group 2')`.

Are there any plans to remedy this within the library itself, or is it something users will need to be aware of when writing custom specifications?

``` php
class GroupSpec implements RulerZ\Spec\Specification
{
    private $group;

    public function __construct($group)
    {
        $this->group = $group;
    }

    public function getRule()
    {
        return 'group = :group';
    }

    public function getParameters()
    {
        return [
            'group' => $this->group
        ];
    }
}


$spec = new RulerZ\Spec\OrX([
    new GroupSpec('group 1'),
    new GroupSpec('group 2')
]);

var_dump($spec->getParameters());
/*
 * array(1) {
 *  'group' =>
 *  string(7) "group 2"
 * }
 */

var_dump($spec->getRule());
// string(32) "group = :group OR group = :group"
```

**Comment by estelsmith** [comment]
With the issue posted, I think that I may be able to create a specification that utilizes the "in" operator from Hoa\Ruler, but there are some cases where it doesn't make much sense, such as when it would be more sensible to say 'x or y'.

**Comment by K-Phoen** [comment]
Yup, I also ran across this limitation but I'm not sure how to fix it or even if I should fix it.
Doctrine has the same issue for instance.

**Comment by estelsmith** [comment]
After playing with the use case in the original example, I figured the method below isn't too much of a compromise.

However, it still leaves open the limitation of not being able to use that particular specification multiple times.

``` php
class InAnyGroups implements RulerZ\Spec\Specification
{
    private $groups;

    public function __construct(array $groups)
    {
        $this->groups = $groups;
    }

    public function getRule()
    {
        return 'group in :groups';
    }

    public function getParameters()
    {
        return [
            'groups' => $this->groups
        ];
    }
}

$spec = new InAnyGroups([
    'group 1',
    'group 2'
]);

$data = [
    [
        'name' => 'test 1',
        'group' => 'group 1'
    ],
    [
        'name' => 'test 2',
        'group' => 'group 2'
    ],
    [
        'name' => 'test 3',
        'group' => 'group 3'
    ]
];

$ruler = new RulerZ\RulerZ(
    new RulerZ\Interpreter\HoaInterpreter(),
    [
        new RulerZ\Executor\ArrayExecutor()
    ]
);

var_dump($ruler->filterSpec($data, $spec));
```

There is one method to remedying the parameter collisions by using positional parameters like Doctrine ORM does when it builds queries itself (like EntityRepository::findOneBy()).

Alternatively, if named parameters are to be continued, maybe collisions could be detected, then remedied by appending a number to the end of it. For instance, if "group" already existed when Spec\Composite is merging parameters together, it could change the parameter name to "group1", "group2", etc.

The downside to this is modifying the returned rule to update its parameter reference, which could be somewhat difficult unless parsing were involved.

I'm personally leaning towards positional parameters, because detecting and remedying collisions doesn't seem so straightforward.

Thoughts?

**Comment by estelsmith** [comment]
With the collision detection, I'd imagine the output of the getRules() and getParameters() methods to look something like this:

```php
$spec = new RulerZ\Spec\OrX([
    new GroupSpec('group 1'),
    new GroupSpec('group 2')
]);

var_dump($spec->getParameters());
/*
 * array(2) {
 *  'group' =>
 *  string(7) "group 1"
 *  'group1' =>
 *  string(7) "group 2"
 * }
 */

var_dump($spec->getRule());
// string(32) "group = :group OR group = :group1"
```

**Comment by estelsmith** [comment]
With positional parameters, I'd imagine it to look a bit like this:

``` php
class GroupSpec implements RulerZ\Spec\Specification
{
    private $group;

    public function __construct($group)
    {
        $this->group = $group;
    }

    public function getRule()
    {
        return 'group = ?';
    }

    public function getParameters()
    {
        return [$this->group];
    }
}

$spec = new RulerZ\Spec\OrX([
    new GroupSpec('group 1'),
    new GroupSpec('group 2')
]);

var_dump($spec->getParameters());
/*
 * array(2) {
 *   [0] =>
 *   string(7) "group 1"
 *   [1] =>
 *   string(7) "group 2"
 * }
 */

var_dump($spec->getRule());
// string(22) "group = ? OR group = ?"
```

**Comment by K-Phoen** [comment]
Introducing positional parameters won't really solve the problem. Even worse, it would make it more difficult to handle parameters: what would happen if we combine (using OrX/AndX) specifications mixing positional and named parameters?

Right now, the only "implementable" solution I can think of is collisions detection. The specifications objects shouldn't be responsible for parameters aliasing (and even less for rewriting rules), and the RulerZ engine can't do it itself as parameters are already merged when it receive them.

To address this issue, I implemented collisions detection at the AndX and OrX level.

WDYT?

**Comment by estelsmith** [comment]
Collision detection is nice, and lets users know when they've stepped into this trap. When I mentioned positional parameters, I was really thinking of ONLY positional parameters being supported, for the very reason you mentioned about mixing positional and named parameters. If we allowed named parameters without collision detection at all, we'd still be in the same boat.

How I've been getting around collisions in RulerZ is by assigning a unique name to each parameter for every instance of the specification.

Unfortunately, while collision detection at least addresses the issue, I do not think it's proper to put the onus of avoiding collisions onto the user instead of being handled by the library itself. I still think the only way to flat-out avoid collisions is to use positional parameters.

Example:

``` php
class GroupSpec implements RulerZ\Spec\Specification
{
    private $group;

    private $parameterName;

    public function __construct($group)
    {
        $this->group = $group;
        $this->parameterName = uniqid();
    }

    public function getRule()
    {
        return sprintf('group = :%s', $this->parameterName);
    }

    public function getParameters()
    {
        return [
            $this->parameterName => $this->group
        ];
    }
}
```

Do you have any other thoughts regarding this issue?

**Comment by K-Phoen** [comment]
I agree that letting the end-user deal with this issue is the wrong solution. Positional parameters could solve the issue but it comes with a cost: the readability. I tend to think that named parameters are easier to read than positional ones.

I definitely agree with you though: this issue needs to be addressed.

**Comment by estelsmith** [comment]
While positional parameters do constitute a loss of readability, I imagine the loss is very slight.

Take `group = ? OR group = ?` into consideration. It's just as readable as `group = :group1 OR group = :group2`. For much larger specifications, I do realize that it becomes more taxing to read the final spec, but I still think it's rather doable.

So while I agree that there is a slight loss of readability, I think the benefit of avoiding collisions is worth it.

Then again, I may have simply grown accustomed to reading the generated queries from Doctrine.

**Comment by estelsmith** [comment]
Positional Pros         | Named Pros
----------------------------|-------------------
- Avoids Collisions    | - Easier Readability

**Comment by K-Phoen** [comment]
> So while I agree that there is a slight loss of readability, I think the benefit of avoiding collisions is worth it.

You're totally right. I'll removed named parameters in favor of positional ones in the next days (our hours if I can).

**Comment by estelsmith** [comment]
I would preferably like to be able to support named and positional parameters at the same time, while allowing them to co-exist. Just in case users mix them for whatever reason I can't think of right now.

I hacked around a bit, and came up with the following: estelsmith/rulerz@2819ac6fa76675bbdb8c8a22e56c21657089786e

It just so happens to allow positional parameters, named parameters, and a mixture of both.

As example of mixing positional and named:

``` php
$data = [
    [
        'name' => 'test 1',
        'group' => 'group 1'
    ],
    [
        'name' => 'test 2',
        'group' => 'group 2'
    ],
    [
        'name' => 'test 3',
        'group' => 'group 3'
    ]
];

$rule = 'group = :my_group OR group = ?';

var_dump($ruler->filter($data, $spec, [
    'my_group' => 'group 1',
    'group 3'
]));
/*
 * array(2) {
 *   [0] =>
 *   array(2) {
 *     'name' =>
 *     string(6) "test 1"
 *     'group' =>
 *     string(7) "group 1"
 *   }
 *   [2] =>
 *   array(2) {
 *     'name' =>
 *     string(6) "test 3"
 *     'group' =>
 *     string(7) "group 3"
 *   }
 * }
 */
```

Basically, `RulerZ::translatePositionalParameters()` will take `group = :my_group OR group = ?` and translate it to `group = :my_group OR group = :0`. Then, in the case of arrays, `ArrayExecutor::prepareParameters()` will turn numeric indexes into indexes prefixed with a colon, turning 0 into ':0'.

The same type of translation that happens in `ArrayExecutor::prepareParameters()` will need to happen in the other executors, or perhaps be moved up into RulerZ itself so executors don't need to bother.

Thoughts regarding this approach?

**Comment by estelsmith** [comment]
To counter my own comment, perhaps it would be best to make the positional parameters part of the query language itself rather than a string-based hack prior to being parsed into the AST.

**Comment by K-Phoen** [comment]
I also think that parameters (positional or named ones) should be part of the grammar. I started [a branch](https://github.com/K-Phoen/rulerz/tree/parameters) in which I updated the grammar and implemented positional parameters mixed with named ones (still WIP).

Feedbacks appreciated :)

**Comment by estelsmith** [comment]
Overall I like the idea. I haven't gotten to play with it yet, though. Although I did make a small comment regarding `HoaVisitor::visit()`.

**Comment by K-Phoen** [comment]
As the _parameters_ branch has been merged, I close this issue.

**Comment by nicholasruunu** [comment]
> Example:
> 
> ```
> class GroupSpec implements RulerZ\Spec\Specification
> {
>     private $group;
> 
>     private $parameterName;
> 
>     public function __construct($group)
>     {
>         $this->group = $group;
>         $this->parameterName = uniqid();
>     }
> 
>     public function getRule()
>     {
>         return sprintf('group = :%s', $this->parameterName);
>     }
> 
>     public function getParameters()
>     {
>         return [
>             $this->parameterName => $this->group
>         ];
>     }
> }
> ```
>

Since this issue is referenced in the code, I'd just like to comment that using `uniqid()` will make the compiled executor totally uncachable.

It's better to use `$this->parameterName = 'group_' . md5(__CLASS__ . $group)` to at least cache the executor per group. (`group_` is just for readability)

Best thing would be to only use positional parameters right now, but that's not documented and mixing with named parameters doesn't seem to work at least with the Doctrine target.

**Comment by estelsmith** [comment]
@nicholasruunu Wasn't this issue resolved back in 2015?

**Comment by nicholasruunu** [comment]
@estelsmith In some sense it probably is, but this issue is still [referenced in the code](https://github.com/K-Phoen/rulerz/blob/master/src/Spec/Composite.php#L82) and we were having issues applying a positional parameterised spec to an already existing doctrine query with named parameters.

It's probably a problem with Doctrine or with the Doctrine target, but we had some caching issues with this previous solution so it led me here.

**Comment by cedvan** [comment]
Hi, 

Why not prefix the parameters names automatically with the object identifier (oid) of the specification to which it belongs?

A specification in my opinion must be autonomous, even when it is included in a chain, and therefore the dynamic parameters configured for a specification must belong only to the context of that specification and not to the context of another.

I may have misunderstood and I may be on the wrong track?
